### ISSUE TITLE
PIXI v5 support

### ISSUE THREAD

**Issue by serglider** [pixijs/pixi-animate#95]
Is there a plan to fully support PIXI v5?

**Comment by olehs1** [comment]
@bigtimebuddy

**Comment by herrKlein** [comment]
Don't know why no one replied, but I'm happy to notice there is a V5 rewrite in this branch https://github.com/pixijs/pixi-animate/tree/ts-and-v5-rewrite

**Comment by bigtimebuddy** [comment]
The v5 rewrite is a WIP by @andrewstart, needs some attention but we’ll get there soon. If anyone wants to help, we’d love the help.

**Comment by andrewstart** [comment]
Very much a WIP. Currently converting to Typescript, then will work on V5
support, without dropping V4 support if possible. Some work will likely
need to happen to update output from the plugin.

**Comment by kulabros** [comment]
Hi there! I'm trying to get work update of my project libraries, migrating from pixi v4 to v5. I'm also using many plugins - pixi-animate, pixi-spine, pixi-filters,.. And I'm I missing something. or.. Pixi 5.3.3., tried both releases versions of pixi-animate and also 2.0.0-rc5, I'm on ES6, webpack, using assets exported from Adobe Animate. I've read release notes of the newest rc, about PIXI namespace and changes, but I'm still not able to get it working. I tried expose PIXI in webpack using ProvidePlugin, setting PIXI on window, I'm importing pixi like 'import * as PIXI from "pixi.js"'..  But I get error on PIXI.animate.MovieClip - PIXI.animate is undefined.. So, I tried to import classes directly 'import {Container, MovieClip} from "pixi-animate"' in assets file.. Then I get Container.extend is not a function - the one that is used in output file from animate cc... So.. I think I'm missing something, some incopability issue maybe, or is it issue with pixi global namespace? Thanks a lot in advance for any help.

**Comment by andrewstart** [comment]
In order to use your assets as is with 2.0.0-rc5, you'll need to use the legacy bundle (in addition to making sure that `PIXI` is a global), which is imported from the file `pixi-animate/dist/pixi-animate-legacy.js`. See https://github.com/pixijs/pixi-animate/tree/ts-and-v5-rewrite#breaking-changes-from-v13

Alternatively, you can update your assets to the 2.0.0 format, and while you'll have to change how you pull in your assets, you won't need a `PIXI` global variable. See https://github.com/pixijs/pixi-animate/tree/ts-and-v5-rewrite#asset-changes

**Comment by kulabros** [comment]
Awesome! I've missed this, I tried to add legacy bundle and it helped, a lot, now I'm successfully running my app on pixi v5.3.3! Thank you very much for the help, next time I'll try to read better the migration changes.

**Comment by pawelstan** [comment]
@kulabros would you share how you imported the pixi-animate-legacy.js to your project ?

**Comment by kulabros** [comment]
> @kulabros would you share how you imported the pixi-animate-legacy.js to your project ?

Sorry, but I haven't catched the message on time. The project libs update was delayed, now I get back to it again and I'm trying to figure out how to run on v6 with older assets support. Unfortunately, I've lost the working example of update I did at the time and I don't remember how I solve it. No I'm trying to update project too and I have no idea yet how to figure out the global namespace for pixi, and pixi-animate-legacy lib using webpack.. I tried what is written in the docs, but it's not working for me. Now I'm trying to run newest versions of libs - pixi.js v6.0.4, pixi-animate 2.0.0-rc6.

Example  of my previous (v4) usage in the webpack module:
I have file called pixi.js > src/js/vendor/pixi.js where all pixi imports are made the same as PIXI expose to the window and it is exported, so everywhere in other modules where I need pixi I just did: import PIXI from "vendor/pixi"

No I tried to change it for v6, so I changed pixi.js import to:
`import * as PIXI from 'pixi.js';
tried window.PIXI = PIXI;`

than import "pixi-animate" ... and as I need to support older assets, I tried to import "pixi-animate/dist/pixi-animate-legacy" just it self or in addition after importing "pixi-animate" but it keeps saying reference error PIXI is not defined from the pixi-animate-legacy, even when the expose to the window is before the import. 

New way of how I try to get it working is:
`import * as PIXI from 'pixi.js';

window.PIXI = PIXI;

import "pixi-animate";
import "pixi-animate/dist/pixi-animate-legacy.js"


export default PIXI;`

But I get to the error:
Uncaught ReferenceError: PIXI is not defined ... at Object../node_modules/pixi-animate/dist/pixi-animate-legacy.js

So I'm still missing something. Do I need to add provide plugin to the webpack, or set the pixi global namespace some new way in the webpack? I'm running webpack 4.44.2, or how should I fix the PIXI namespace, so it is accessible for the pixi-animate? I guess I'm missing somthing.. Kind regards for any help.

**Comment by kulabros** [comment]
> In order to use your assets as is with 2.0.0-rc5, you'll need to use the legacy bundle (in addition to making sure that `PIXI` is a global), which is imported from the file `pixi-animate/dist/pixi-animate-legacy.js`. See https://github.com/pixijs/pixi-animate/tree/ts-and-v5-rewrite#breaking-changes-from-v13
> 
> Alternatively, you can update your assets to the 2.0.0 format, and while you'll have to change how you pull in your assets, you won't need a `PIXI` global variable. See https://github.com/pixijs/pixi-animate/tree/ts-and-v5-rewrite#asset-changes

Hi Andrew, replying almost after year, so things have had a changed a bit, new versions are released.. As I wrote in my previous reply, I'm trying to migrate my project from pixi v4 > v6, and I'm using animate and spine plugins and I have assets in older format, and I tried to convert the assets, but there are some differences (stage definitions are removed) so the script won't convert it and ends in error, so I think about to keep the assets as they are and add the legacy support, but I have some problems according to the global pixi namespace for the animate, is there some example, jsfiddle or any docs regarding to usage of pixi-animate under webpack for the latest versions? What I've founf in the documentation have not helped me, I think it is some issue related to the webpack version / pixi plugin global namespace defition... Thanks a lot for any help.

**Comment by andrewstart** [comment]
Could you post one of the assets that doesn't convert so that I can see what I need to do to fix the conversion script? I'll investigate webpack docs, but I am guessing the solution to using the legacy build is to configure webpack to assign `pixi.js` to the `PIXI` variable, or something like that.

**Comment by kulabros** [comment]
Yes, here's one example. We use it only as i library, we're not using objects on stage, so there's no stage objects definition. [https://www.dropbox.com/s/le2d4ukshjh07no/Multiplier.js?dl=0](url)

**Comment by kulabros** [comment]
Amazing, thank you very much indeed, I'll give it a try! And thanks for clarifying that pixi-animate-legacy is complete build, I though so as I saw the code, but "...you'll need to use the legacy bundle (in addition..." I've only just understand it wrong. I'll leave comment later on about how I moved. Thank you very much again for the help!

**Comment by kulabros** [comment]
> I've made a new release candidate that allows for your file to be converted, with a warning that it doesn't know about assets and what you need to do to fix that (`myAssetData.getTexture = PIXI.Texture.fromFrame`, for example).
> 
> To go down the legacy build route, you could import PixiJS's IIFE build (which is `pixi.js/dist/pixi.js`), or use [Webpacks Provide Plugin](https://webpack.js.org/plugins/provide-plugin/) to define a PIXI global. Note that `pixi-animate/dist/pixi-animate-legacy.js` is a complete build, and does not require you to import `pixi-animate` separately.



> I've made a new release candidate that allows for your file to be converted, with a warning that it doesn't know about assets and what you need to do to fix that (`myAssetData.getTexture = PIXI.Texture.fromFrame`, for example).
> 
> To go down the legacy build route, you could import PixiJS's IIFE build (which is `pixi.js/dist/pixi.js`), or use [Webpacks Provide Plugin](https://webpack.js.org/plugins/provide-plugin/) to define a PIXI global. Note that `pixi-animate/dist/pixi-animate-legacy.js` is a complete build, and does not require you to import `pixi-animate` separately.

Very cool, helped, now I've successfully converted the assets to the new format. I couldn't use the pixi-animate-upgrade command, but when I run the assetConversion.jsj script manually, now it converts ok, just with the warning.

Awesome, thank you, thank you:)

Along the update of all used plugins in my project, now I'm facing last two issues, but it's for another topic in here, but just for information...
pixi-timer - there are some problems according to the PIXI global namespace, I'll leave there some issue about the problem too
pixi-spine - similar to problem in here - assets in older version 3.7.94.. and there is error from where I use PIXI.spine.core modules Uncaught TypeError: Cannot read property 'core' of undefined ... So, I'll have to solve this issue too, but I know it's a different topic than this ticker.

But again, thanks a lot, your patched helped me to convert the assets and build the application. After I solve timer and spine issues, I'll see how it's working or what's going on - thanks for the tip of alternation getTexture method.

**Comment by kulabros** [comment]
Hello there, I'm allready successfully running my application at the released 5.3.10 version with pixi-animate 2.0.0.-rc.11 - I've migrated the assets to new format and had to solve some issues or mostly differences.. First - I've discovered, I need to call the setup method at the assets, of course. Second, thank to @andrewstart with the help to change the getTexture method I was able to run the animations. Third, I found the migration info notes at the 2.0.0-rc.1 about the shapes and changed  shorthand draw command from 'c' to 'cp', that solved me running shapes and masks. The shapes I've had to deliver to the animate data structure and the same place where getTexture is modified, so I've delivered the shapes external json data, cool, that solved the animations with the shapes. And last thing what is not working as expected it the shapes color - all created shapes are not having color they should have, they are all just black. I love black color, it's my most favorite :) But, I'm missing something another to be changed in the draw commands, or this should work ok - like coloring the shapes.. I'm gonna continue tomorrow to discover more about this or maybe do some fiddle example that it will confirm, but first I've thought aboiut to ask you, maybe you tell me I'm doing wrong using the shapes. yeah but it's another story, I'm migrating allready created project. Thank you very much for any help about the black shapes. have a nice day and good luck!

**Comment by andrewstart** [comment]
Hmm, I don't have any thoughts on why shapes would be all black, beyond the fact that the color probably isn't decoding properly when it is read from the shapes file. Seeing a fiddle with such an asset would help.

**Comment by pawelstan** [comment]
You need to Convert all hex string colors (animate) to int (pixi.js). Below part of my loader.use

function symbolLoader(resource, next) {
	if (resource.url.search(/\.shapes\.(json|txt)$/i) > -1) {
		const items = window.PIXI["animate"].ShapesCache[resource.name] = window.PIXI["animate"].utils.deserializeShapes(resource.data);

		// Convert all hex string colors (animate) to int (pixi.js)
		for (let i = 0; i < items.length; i++)
		{
			const item = items[i];
			for (let j = 0; j < item.length; j++)
			{
				const arg = item[j];
				if (typeof arg === "string" && arg[0] === "#")
				{
					item[j] = utils.hexToUint(arg);
				}
			}
		}
	}
	next();
}

**Comment by andrewstart** [comment]
Additionally, anything that is getting loaded via the library's `load()` method should be having that handled already.

**Comment by kulabros** [comment]
Thank you both @andrewstart @pawelstan  for replies. Actually, I've found, that only thing I had to change that fixed the colors was re-enabling commented code where the shapes are added to the shapes cache, after calling ShapesCache.add(shapes[0].id, shapes[0].json), and keeping the code which adds the imported json (webpack) shapes to the animate cc animations object under shapes property, and to keep the functionality for the converted format, I'm adding the shapes under it's name id property - example: 
`
let animations = ‘imported animate cc animations.js by webpack’;
let shapesJSON = ímported shapes json exported by animate cc by webpack’;
let shapes = [name ids if shapes jsons added to the ShapesCache];

animations = animations.animations.setup(Animate);

if (animations.getTexture) {
	animations.getTexture = PIXI.Texture.from;
}

if (shapes.lenght > 0 && animations.shapes) {
	const shapesNameId = shapesJSON[0].id;
	animations.shapes[shapesNameId] = shapesJSON[0].json;
}
`
And now it's working as with the previous version of assets and pixi and pixi-animate libs. great, thanks a lot for help. Adding the shapes to the shapes cache solved the issue that all shapes were black, no additional coding was needed in my case, but abyway, thanks a lot for the code and effort! Have a nice day :)
